man/authd_setup.8
.\" Copyright Georgios Portokalidis, 2004.
.\"
.\" This file may be copied under the conditions described
.\" in the GNU GENERAL PUBLIC LICENSE, Version 2, June 1991
.\" that should have been distributed together with this file.
.\"
.TH AUTHD_SETUP 8 "May 28, 2004" "Version 0.8" "System Manager's Manual"
.SH NAME
authd_setup.sh \- Admission control setup script
.SH SYNOPSIS
.B authd_setup.sh
.IB install "|" uninstall "|" policy "|" keys "|" credentials 
.SH DESCRIPTION
A script to help setting up a system for admission control. It creates default
files needed by authd(8) for operation and furthermore can be used to generate
credentials.
.br
Clients can also use it to generate key pairs.
.P
For more fine grained control keynote(1) should be used directly to generate
keys. Note that client keys need to be RSA.
.P
On how to write policy files consult keynote(5).
.P
Note that this script cannot be used to initialise resource control. Please read
\'doc/RESOURCE_CONTROL.txt\' and authdb_manage(8) for more information. 
.SH OPTIONS
.TP
.RI "" install " [ " user_id " ] [ " pathname " ]" 
.RI "Install all necessary files for admission control in " pathname " using
.RI "user id " user_id ". Defaults are \'/etc/authd' and the active user id."
Files installed include public, private key and a policy. Additionally the file
\'/tmp/\.authd\' is created to be used for IPC communication. If you wish to
use another file you can, by specifying it when running authd.
.TP
.RI "" uninstall " [ " pathname " ] "
.RI "Uninstall the files installed in " pathname ". It will also remove the
directory if it is empty. \'/tmp/.authd\' will also be removed.
.TP
.RI "" policy " [ " user_id " ] [ " pathname " ]"
.RI "Create a default policy file, install it in " pathname " and set 
.RI "ownership to " user_id "."
.TP
.RI "" keys " [ " key_size " ] [ " pubk_name " ] [ " privk_name " ]" 
.RI "Create an RSA key pair of " key_size " bits and store the public key to
.RI "file " pubk_name " and the private key to " privk_name "."
Default key size is 1024 bits and default filenames are \'pub.key\' and
\'priv.key\'.
.TP
.RI "" credentials " " pubkey " " conditions " " output " [ " path " ]"
.RI "Create and sign credentials storing them in file specified by " output "."
The licensee of the credentials is the public key contained in file
.I pubkey
.RI "and the conditions used are from file " conditions "."
.RI "If " path " is specified then admission control's keys are looked up there
instead of the default location \'/etc/authd\'.
.SH EXIT STATUS
Zero if terminated successfully, or non-zero on error.
.SH FILES
/etc/authd/authd_key.pub, /etc/authd/authd_key.priv, /etc/authd/policy, 
tmp/.authd
.SH EXAMPLES
.B "authd_setup.sh install authd"
.br
Install files in standard location \'/etc/authd\' and set ownership to user
'authd'.
.P
.B "authd_setup.sh keys 2046 mypub mypriv"
.br
Create a key pair of 2046 bits size and store public key to \'mypub\' and
private key to \'mypriv\'.
.SH SEE ALSO
authd(8), authdb_manage(8), keynote(1), keynote(5)
.SH AUTHOR
Georgios Portokalidis (gportoka@liacs.nl)



ChangeLog
ADMCTRL - Admission Control Daemon - ChangeLog
--------------------------------------------

0.8.9
=====

Configuration
  * configure.ac Fixed for additional compatibility
  * src/Makefile.am Updated to reflect changes in configure.ac
  * tests/Makefile.am Updated to reflect changes in configure.ac
  * kernel/kadmctrl_req.c Removed the 'k' prefix from functions
  * kernel/kclient.c Updated to reflect changes in kernel/kadmctrl_req.c

0.8.8
=====

Configuration
  * configure.ac Added options to specify seperate keynote header and 
  library directory,

Tests
  * tests/enc_nonce Encrypts a nonce and prints it out in hex.
  * kernel/kclient.c Kernel driver test client.

Kernel Driver
  * kernel/kadmctrl_req.c Admission control request manipulation function
  for the kernel.

0.8.7
=====

Resource control
  * arith_parser.c updated postfix_expr_parse() & infix_expr_parse() to
  successfully parse an empty string and return 0 as a result.

0.8.6
=====

Documentation
  Added authd_setup(8) man page.
  Updated authd(3) man page.

Admission control client library
  Function admctrl_req_set_authinfo() now accepts buffers
  containing the public key and credentials instead of the files containing 
  them.
  Removed function admctrl_req_encrypt_nonce(). It's an end-user assisting
  function. Code provided as a C snippet in 'doc/encrypt_nonce.c'.

0.8.5
====

Documentation
  Added more documentation under doc/.
  Written tests/README.

Configuration
  Corrected configure.ac, src/Makefile.am and tests/Makefile.am to link only
  necessary libraries.

Tests
  Updated client to support resource control and test function type arguments.

0.8.4
=====

Documentation
  Added authd(3) man page for developers and updated documentation in doc/.

Configuration
  Added option --disable-authd to enable stand alone compilation of the client
  library.

0.8.3 
=====

Admission control front-end
  Added support for linux kernel module device driver.
  Renamed front-end executable to 'authdfe'.

Configuration
  Changed --without-clientlib to --disable-clientlib.
  Changed --with-frontend to --enable-front-end.
  Changed --with-keynote-prefix to --with-keynote.
  Changed --with-db-prefix to --with-berkeley-db.
  Changed --with-pthread-prefix to --with-pthread.
  Changed --with-snprintfv-prefix to --with-snprintfv.

0.8.2
=====

Added a linux kernel module device driver

0.8
===

Started using the GNU build system


authd_setup.sh
#!/bin/bash

DIR=/etc/authd
TMPFILE=/tmp/.authd
CLPUB=pub.key
CLPRIV=priv.key
CLKEYSZ=1024
RESCRTL=resourcectrl

generate_policy()
{
  echo "KeyNote-Version: 2" >$POLICY
  echo "Authorizer: \"POLICY\"" >>$POLICY
  echo -n "Licensees: " >>$POLICY
  cat $PUBKEY >>$POLICY
  echo "Conditions: app_domain == \"MY DOMAIN\" -> \"true\";" >> $POLICY
}

case "$1" in
  install)
		if [ "$#" -lt "2" ]; then
			INSTALLAS=$USER
		else
			INSTALLAS=$2
		fi
    if [ "$#" -gt "2" ]; then
      DIR=$3
    fi
    PUBKEY=$DIR/authd_key.pub
    PRIVKEY=$DIR/authd_key.priv
    POLICY=$DIR/policy

    echo "Creating admission control files in $DIR"
    mkdir -m 755 -p $DIR/$RESCRTL
    echo "Generating admission control keys..."
    keynote keygen rsa-base64: $CLKEYSZ $PUBKEY $PRIVKEY
    chmod 644 $PUBKEY
    chmod 600 $PRIVKEY
    echo "Generating default policy file..."
    generate_policy
    echo "Creating $TMPFILE file for IPC..."
    touch $TMPFILE
    chmod 644 $TMPFILE
    echo "Setting ownership of files to $INSTALLAS..."
    chown -R $INSTALLAS $DIR $TMPFILE
    exit 0
    ;;
  uninstall)
    if [ "$#" -gt "2" ]; then
      DIR=$2
    fi
    PUBKEY=$DIR/authd_key.pub
    PRIVKEY=$DIR/authd_key.priv
    POLICY=$DIR/policy

    echo "Removing admission control files located in $DIR..."
    rm -f $PUBKEY $PRIVKEY $POLICY
    rmdir --ignore-fail-on-non-empty $DIR/$RESCRTL
    rmdir --ignore-fail-on-non-empty $DIR
    echo "Removing $TMPFILE file..."
    rm -f $TMPFILE
    exit 0
    ;;
  keys)
    if [ "$#" -gt "1" ]; then
      CLKEYSZ=$2
    fi
    if [ "$#" -gt "2" ]; then
      CLPUB=$3
    fi
    if [ "$#" -gt "3" ]; then
      CLPRIV=$4
    fi
    echo "Creating key pair..."
    rm -f $CLPUB $CLPRIV
    keynote keygen rsa-base64: $CLKEYSZ $CLPUB $CLPRIV
    exit 0
    ;;
  policy)
		if [ "$#" -lt "2" ]; then
			INSTALLAS=$USER
		else
			INSTALLAS=$2
		fi
    if [ "$#" -gt "2" ]; then
      DIR=$3
    fi
    PUBKEY=$DIR/authd_key.pub
    POLICY=$DIR/policy

    mkdir -m 755 -p $DIR
    echo "Generating default policy file..."
    generate_policy
    exit 0
    ;;
  credentials)
    if [ "$#" -lt "4" ] ; then
      echo "Usage: $0 credentials public_key conditions output [path]"
      exit 1
    fi
    if [ "$#" -gt "4" ]; then
      DIR=$5
    fi
    PUBKEY=$DIR/authd_key.pub
    PRIVKEY=$DIR/authd_key.priv

    if  [[ ! -r "$2" || ! -r  "$3" || ! -r "$PUBKEY" || ! -r "$PRIVKEY" ]] ; then
      echo "Check that the files $2,$3,$PUBKEY,$PRIVKEY exist and you have permissions to read them"
      exit 1
    fi

		TMP=`mktemp creds-XXXXXX` || exit 1
    rm -f $TMP 
    echo "KeyNote-Version: 2" >$TMP
    echo -n "Authorizer: " >>$TMP
    cat $PUBKEY >>$TMP
    echo -n "Licensees: " >>$TMP
    cat $2 >>$TMP
    echo -n "Conditions: " >> $TMP
    cat $3 >> $TMP
    echo -n "Signature: " >>$TMP
    rm -f $4
    cat $TMP > $4
    keynote sign "sig-rsa-sha1-base64:" $TMP $PRIVKEY >> $4
    rm -f $TMP
    exit 0
    ;;
  *)
    echo "Usage: $0 install|uninstall|policy|keys|credentials"
		echo "  install     [user id] [pathname]"
    echo "     Install all necessary files for admission control in 'pathname'"
    echo "     and set ownership to 'user id'.pathname is '/etc/authd' by default."
		echo "  uninstall   [pathname]"
    echo "     Remove files installed by install in 'pathname'."
		echo "  policy      [user id] [pathname]"
    echo "     Create a new default policy file in 'pathname' and set ownership"
    echo "     to 'user id'."
		echo "  keys      [key size] [pubk name] [privk name]"
    echo "     Generate a public/private key pair. The length of "
    echo "     of the key will be 'key size', default 1024 bits. Write public"
    echo "     key to file 'pubk name' and private to 'privk name', default "
    echo "     'pub.key' and 'priv.key'"
		echo "  credentials (public key) (conditions) (output) [path]"
    echo "     Generate credentials for licensee in 'public key' using"
    echo "     'conditions' and store it to file 'output'. Optionally 'path'"
    echo "     is the location of admission control keys."
		echo ""
    exit 1
    ;;
esac

exit 0
